code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/filter/snapshot/SnapshotFilter.java

```diff
@@ -0,0 +1,81 @@
+import io.cattle.platform.iaas.api.filter.common.AbstractDefaultResourceManagerFilter;
+import io.cattle.platform.object.ObjectManager;
+import io.cattle.platform.object.util.DataUtils;
+import io.github.ibuildthecloud.gdapi.exception.ValidationErrorException;
+import io.github.ibuildthecloud.gdapi.request.ApiRequest;
+import io.github.ibuildthecloud.gdapi.request.resource.ResourceManager;
+import io.github.ibuildthecloud.gdapi.validation.ValidationErrorCodes;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SnapshotFilter extends AbstractDefaultResourceManagerFilter {
+
+    private static final Logger log = LoggerFactory
+            .getLogger(SnapshotFilter.class);
+    ObjectManager objectManager;
+
+    @Override
+    public String[] getTypes() {
+        return new String[] { "snapshot" };
+    }
+
+    @Override
+    public Class<?>[] getTypeClasses() {
+        return new Class<?>[] { Snapshot.class };
+    }
+
+    @Inject
+    public void setObjectManager(ObjectManager objectManager) {
+        this.objectManager = objectManager;
+    }
+
+    @Override
+    public Object create(String type, ApiRequest request, ResourceManager next) {
+        Long volumeId = DataUtils.getFieldFromRequest(request,
+                SnapshotConstants.FIELD_VOLUME_ID, Long.class);
+        if (volumeId == null) {
+            throw new ValidationErrorException(
+                    ValidationErrorCodes.MISSING_REQUIRED,
+                    SnapshotConstants.FIELD_VOLUME_ID);
+        }
+
+        Volume volume = objectManager.loadResource(Volume.class, volumeId);
+        if (volume == null) {
+            throw new ValidationErrorException(
+                    ValidationErrorCodes.MISSING_REQUIRED,
+                    SnapshotConstants.FIELD_VOLUME_ID);
+        } else if (volume.getImageId() != null) {
+            throw new ValidationErrorException(
+                    ValidationErrorCodes.INVALID_ACTION,
+                    SnapshotConstants.FIELD_VOLUME_ID);
+        } else if (!volume.getUri().contains(SnapshotConstants.MANAGED_TAG)) {
+            throw new ValidationErrorException(
+                    ValidationErrorCodes.INVALID_ACTION,
+                    SnapshotConstants.FIELD_VOLUME_ID);
+        }
+        String uri = volume.getUri();
+        String volumePath = uri.split(SnapshotConstants.MANAGED_TAG)[1];
+        // Volume would be in format of <user>/<volume_name>/<volume_uuid>
+        String[] volumeParameters = volumePath.split("/");
+        if (volumeParameters.length != 3) {
+            log.error("Error with parsing volume path: " + uri);
+            throw new ValidationErrorException(
+                    ValidationErrorCodes.INVALID_OPTION,
+                    SnapshotConstants.FIELD_VOLUME_ID);
+        }
+
+        Snapshot snapshot = request.proxyRequestObject(Snapshot.class);
+        DataUtils.getWritableFields(snapshot).put(
```

Comment:
When you send the event to the python-agent just send the volume and get the uuid at that time, so you don't hvae to copy the data.